### DISCUSSION TITLE
URL do block with custom downloader strategy

### DISCUSSION THREAD

**Discussion by marble-sh** [Homebrew/discussions#718] [Casks] [answered]
The docs hint at how to do this, but don't provide an example.

https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/url.md#mixing-additional-url-parameters-with-the-block-syntax

> In rare cases, you might need to set URL parameters like cookies or referer while also using the block syntax.

> This is possible by returning a two-element array as a block result. The first element of the array must be the download URL; the second element must be a Hash containing the parameters.

But when I try to do this like so:

```ruby
  url do
    [
      url,
      {using: MyDownloadStrategy}
    ]
  end
```

I get the following error:

```
==> Cask::Installer#install
==> Printing caveats
==> Cask::Installer#fetch
Error: Download failed on Cask 'onivim2' with message: wrong number of arguments (given 2, expected 1)
/usr/local/Homebrew/Library/Homebrew/cask/url.rb:38:in `initialize'
/usr/local/Homebrew/Library/Homebrew/cask/dsl.rb:184:in `new'
/usr/local/Homebrew/Library/Homebrew/cask/dsl.rb:184:in `block (2 levels) in url'
```

I'm sure there's something simple I'm doing wrong here, but I can't for the life of me figure out what.

**Comment by SMillerDev** [comment]
@Homebrew/cask can you lend a hand here?

**Comment by marble-sh** [comment, marked as answer]
The trouble seems to be in this block of code in dls.rb:

https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cask/dsl.rb#L184

```ruby
    def url(*args, **options)
      caller_location = caller_locations[0]

      set_unique_stanza(:url, args.empty? && options.empty? && !block_given?) do
        if block_given?
          LazyObject.new { URL.new(*yield, from_block: true, caller_location: caller_location) }
        else
          URL.new(*args, **options, caller_location: caller_location)
        end
      end
    end
```

`URL.new` inside the LazyObject is missing `**options` in the arguments list; adding options to the argument list allowed me to do 

```ruby
  url using: MyDownloadStrategy do
    some_url
  end
```

If this is actually a bug I'd be happy to open a pull request for this issue.

**Reply by vitorgalvao** [reply]
> If this is actually a bug I'd be happy to open a pull request for this issue.

Might be a bug, yes. Pinging @claui, who first developed the feature.

**Reply by marble-sh** [reply]
I hope it's okay, but I went ahead and made the pull requests, please let me know if I missed something.

* https://github.com/Homebrew/brew/pull/10548
* https://github.com/Homebrew/homebrew-cask/pull/99447
